cmd/admin-revoker/main.go

```diff
@@ -211,7 +240,129 @@ func (r *revoker) revokeByReg(ctx context.Context, regID int64, reasonCode revoc
 
 func (r *revoker) revokeMalformedBySerial(ctx context.Context, serial string, reasonCode revocation.Reason) error {
-	return r.revokeCertificate(ctx, core.Certificate{Serial: serial}, reasonCode)
+	return r.revokeCertificate(ctx, core.Certificate{Serial: serial}, reasonCode, false)
+}
+
+// blockByPrivateKey blocks future issuance for certificates with a public key
+// matching the SPKI hash of the provided private key. Reason code must be 1
+// (Key Compromise). Keys passed to this function will be validated as a pair
+// before any actions are taken. This method does not revoke any certificates
+// directly; however 'bad-key-revoker', which references the 'blockedKeys'
+// table, will eventually revoke certificates with a matching SPKI hash.
+func (r *revoker) blockByPrivateKey(ctx context.Context, privateKey crypto.Signer, reasonCode revocation.Reason) error {
+	if reasonCode < 0 || reasonCode != 1 {
+		panic(fmt.Sprintf("Invalid reason code: %d", reasonCode))
+	}
+
+	err := verifyPrivateKey(privateKey)
+	if err != nil {
+		return err
+	}
+
+	spkiHash, err := getPublicKeySPKIHash(privateKey.Public())
+	if err != nil {
+		return err
+	}
+
+	u, err := user.Current()
+	if err != nil {
+		return err
+	}
+
+	req := &sapb.AddBlockedKeyRequest{
+		KeyHash:   spkiHash,
+		Added:     r.clk.Now().UnixNano(),
+		Source:    "admin-revoker",
+		Comment:   fmt.Sprintf("blocked by %s", u),
+		RevokedBy: 0,
+	}
+
+	_, err = r.sac.AddBlockedKey(ctx, req)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// revokeByPrivateKey revokes all certificates with a public key matching the
+// SPKI hash of the provided private key. Reason code must be 1 (Key
+// Compromise). Keys passed to this function will be validated as a pair before
+// any actions are taken. The provided key will not be added to the
+// 'blockedKeys' table, this is done to avoid a race between 'admin-revoker'
+// and 'bad-key-revoker'.
+func (r *revoker) revokeByPrivateKey(ctx context.Context, privateKey crypto.Signer, reasonCode revocation.Reason) error {
+	if reasonCode < 0 || reasonCode != 1 {
+		panic(fmt.Sprintf("Invalid reason code: %d", reasonCode))
+	}
+
+	err := verifyPrivateKey(privateKey)
+	if err != nil {
+		return err
+	}
+
+	spkiHash, err := getPublicKeySPKIHash(privateKey.Public())
+	if err != nil {
+		return err
+	}
+
+	matches, err := r.getCertsMatchingSPKIHash(spkiHash)
+	if err != nil {
+		return err
+	}
+
+	for i, match := range matches {
+		err := r.revokeBySerial(context.Background(), match, revocation.Reason(reasonCode), true)
+		if err != nil {
+			return fmt.Errorf(
+				"failed to revoke serial %q, entry %d of %d affected certificates: %s",
+				match,
+				(i + 1),
+				len(matches),
+				err,
+			)
+		}
+	}
+	return nil
+}
+
+func (r *revoker) spkiHashInBlockedKeys(spkiHash []byte) (bool, error) {
+	var count int
+	err := r.dbMap.SelectOne(&count, "SELECT COUNT(*) as count FROM blockedKeys WHERE keyHash = ?;", spkiHash)
+	if err != nil {
+		if db.IsNoRows(err) {
```

Comment:
Since we're selecting a COUNT, we will never get a NoRows error. There are two ways to do this, mutually exclusive:

 1. Select a row from blockedKeys (doesn't matter what row, since keyHash is a unique index - there will be at most one response), and check for IsNoRows, OR
 2. `SELECT COUNT`, and check for `count > 0`.